go/table.go

```diff
@@ -2,17 +2,37 @@ package flatbuffers
 
 // Table wraps a byte slice and provides read access to its data.
 //
-// The variable `Pos` indicates the root of the FlatBuffers object therein.
+// The byte slice `_buf` is not quite as it appears: We need to track a position
+// within the buffer (the root of the object that this Table represents), in
+// addition to the bounds of the buffer itself. Instead of using another field,
+// we put the position in the len of the slice, and the actual length in the
+// cap. When re-slicing it, `_buf` can be used as-is, but to avoid confusing
+// clients, we provide the Bytes() method to get the whole buffer.
 type Table struct {
-	Bytes []byte
-	Pos   UOffsetT // Always < 1<<31.
+	_buf []byte
+}
+
+// Init sets up the Table to refer to the object at offset `pos` in the buffer `buf`.
+// `pos` must be < 1<<31.
+func (t *Table) Init(buf []byte, pos UOffsetT) {
+	t._buf = buf[:pos:len(buf)]
+}
+
+// Bytes returns the full buffer that this Table was initialized with.
+func (t Table) Bytes() []byte {
+	return t._buf[:cap(t._buf)]
```

Comment:
Why use `cap`, wouldn't that be unsafe if the buffer is reused?

Reply:
The cap of _buf is the len of the slice that Init was originally given, so it basically returns the same slice. The only difference is that we throw out the value of the original cap, but Table doesn't need that for anything. If a caller wants to reuse the buffer with the original cap, they can just hold on to a reference to it themselves.

Reply:
When I say "reused" I mean during the normal course of using a `Builder` in a hot path, where users are repeatedly calling `Reset`. My understanding, and please correct me if this is not true, is that using `cap` here would mean the users would get bogus data that they didn't write, if they wrote a payload that didn't cause the capacity of the slice to change.

Reply:
I might be missing something, but Builder doesn't use Table at all, right? Table is just for the read path.

This cap is not the cap of the buffer given to Init. I agree that would be bad, because it would expose data past the end of that buffer. But Init sets the cap of _buf to be the len of the buffer it's given. And this reads that value out and creates a slice with the correct len. So nothing ever sees bogus data, no matter how buffers are reused. (Except obviously you can't reuse a buffer for building and also be reading the old structure in it at the same time.)

Reply:
Right, thanks--juggling too many PR reviews here I guess :-)